### PR TITLE
Reef 184 gc

### PR DESCRIPTION
**Pull request by gwsshs22**
REEF-184_GC changes existing NetworkService in GroupComm library to new NetworkService. I have tested some examples, BroadcastREEF and BGCLocal, but could not make sure that it works consistently because there are no sufficient test codes about GroupComm.

### CODE REVIEW

**Comment by gwsshs22** [comment]
Actually the BoradcastREEF example in master branch of apache/reef seems to have a bug.

--- the result of 
 java -cp reef-examples.jar org.apache.reef.examples.group.broadcast.BroadcastREEF -dim 2 -receivers 14

4? 22, 2015 11:56:17 ?? org.apache.reef.util.REEFVersion logVersion
??: REEF Version: 0.11.0-incubating-SNAPSHOT
4? 22, 2015 11:56:17 ?? org.apache.reef.wake.remote.address.HostnameBasedLocalAddressProvider <init>
??: Instantiating HostnameBasedLocalAddressProvider
4? 22, 2015 11:56:18 ?? org.apache.reef.wake.remote.address.HostnameBasedLocalAddressProvider <init>
??: Instantiating HostnameBasedLocalAddressProvider
4? 22, 2015 11:56:18 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver <init>
??: Obtained all communication group
4? 22, 2015 11:56:18 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver <init>
??: Added operators to allCommGroup
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-1-1429714579010'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 1, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-2-1429714579262'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 2, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-3-1429714579329'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 3, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-4-1429714579355'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 4, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-5-1429714579381'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 5, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 6, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-6-1429714579402'}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-7-1429714579417'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 7, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-8-1429714579435'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 8, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-9-1429714579459'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 9, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 10, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-10-1429714579480'}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-11-1429714579501'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 11, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 12, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-12-1429714579525'}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-13-1429714579547'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 13, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-14-1429714579568'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 14, Outstanding requests: Optional:{1}
4? 22, 2015 11:56:19 ?? org.apache.reef.examples.group.broadcast.BroadcastDriver$EvaluatorAllocatedHandler onNext
??: Submitting an id context to AllocatedEvaluator: AllocatedEvaluator{ID='Node-15-1429714579588'}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 15, Outstanding requests: Optional:{0}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.driver.ResourceManager sendRuntimeStatus
??: Allocated: 15, Outstanding requests: Optional:{0}
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@5562f7ec 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@ee15b1f 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@5b6feec0 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@373e38a7 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@7846744d 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@67bb3f4a 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@7e956b74 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@3c5508ab 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@21403f7d 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@3ca6a478 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@6779630e 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@eea44aa 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@4b07bb1 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@33ca0d82 
4? 22, 2015 11:56:19 ?? org.apache.reef.runtime.local.process.ReefRunnableProcessObserver onResourceStatus
??: Sending resource status: org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl@10731e16 
4? 22, 2015 11:56:31 ?? org.apache.reef.runtime.common.driver.DriverStatusManager onError
??: Shutting down the Driver with an exception: 
java.lang.NullPointerException
	at org.apache.reef.io.network.group.impl.driver.TreeTopology.addTaskNode(TreeTopology.java:208)
	at org.apache.reef.io.network.group.impl.driver.TreeTopology.addChild(TreeTopology.java:196)
	at org.apache.reef.io.network.group.impl.driver.TreeTopology.addTask(TreeTopology.java:185)
	at org.apache.reef.io.network.group.impl.driver.CommunicationGroupDriverImpl.addTask(CommunicationGroupDriverImpl.java:233)
	at org.apache.reef.examples.group.broadcast.BroadcastDriver$ContextActiveHandler.onNext(BroadcastDriver.java:231)
	at org.apache.reef.examples.group.broadcast.BroadcastDriver$ContextActiveHandler.onNext(BroadcastDriver.java:182)
	at org.apache.reef.runtime.common.utils.BroadCastEventHandler.onNext(BroadCastEventHandler.java:37)
	at org.apache.reef.util.ExceptionHandlingEventHandler.onNext(ExceptionHandlingEventHandler.java:45)
	at org.apache.reef.runtime.common.utils.DispatchingEStage$1.onNext(DispatchingEStage.java:67)
	at org.apache.reef.runtime.common.utils.DispatchingEStage$1.onNext(DispatchingEStage.java:64)
	at org.apache.reef.wake.impl.ThreadPoolStage$1.run(ThreadPoolStage.java:180)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

And this execution was forcibly closed.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 1]
```diff
@@ -69,7 +62,7 @@
  * <p/>
  * Also starts the NameService & the NetworkService on the driver
  */
-public class GroupCommDriverImpl implements GroupCommServiceDriver {
+public class GroupCommDriverImpl implements GroupCommServiceDriver, NetworkEventHandler<GroupCommunicationMessage> {
```
Comment (taegeonum) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: why don't you create a separate NetworkEventHandler class? Is it not possible? 
If you make a separate class, then we can remove the cyclic dependency. (so can remove InjectionFuture)

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 2]
```diff
@@ -196,26 +173,11 @@ public Configuration getContextConfiguration() {
...
-            GroupCommunicationMessageCodec.class)
-        .bindNamedParameter(NetworkServiceParameters.NetworkServiceHandler.class,
-            GroupCommNetworkHandlerImpl.class)
-        .bindNamedParameter(NetworkServiceParameters.NetworkServiceExceptionHandler.class,
-            ExceptionHandler.class)
-        .bindNamedParameter(NameServerParameters.NameServerAddr.class, nameServiceAddr)
-        .bindNamedParameter(NameServerParameters.NameServerPort.class, Integer.toString(nameServicePort))
-        .bindNamedParameter(NetworkServiceParameters.NetworkServicePort.class, "0").build();
+
+    final Configuration retVal = nsConfBuilderFactory.get().createBuilder(null)
```
Comment (taegeonum) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: `.createBuilder()` would be better

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/task/CommunicationGroupClientImpl.java [hunk 3]
```diff
@@ -75,11 +75,11 @@
   @Inject
   public CommunicationGroupClientImpl(@Parameter(CommunicationGroupName.class) final String groupName,
                                       @Parameter(TaskConfigurationOptions.Identifier.class) final String taskId,
-                                      @Parameter(DriverIdentifier.class) final String driverId,
+                                      @Parameter(NetworkServiceParameter.DriverNetworkServiceIdentifier.class) final String driverId,
```
Comment (taegeonum) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/task/CommunicationGroupClientImpl.java]: why not DriverIdentifier?

**Comment by taegeonum** [comment]
@gwsshs22 I've added some comments. Please look at them.

**Pull request closed without merging**